### PR TITLE
Add OpenBSM service to collect socket events on macOS

### PR DESCRIPTION
**Pull request by Wajihulhassan**
Currently, Zeek-agent uses Endpoint Security framework service to collect file and process events on macOS; however, the Endpoint Security framework does not support socket events. Thus, to collect `connect` and `bind` socket events, this PR adds OpenBSM service. 

To use this service, the user only needs to add `nt` flag in `/etc/security/audit_control` configuration file of OpenBSM. After that socket events can be read from `/dev/auditpipe`. This PR takes care of reading events from that pipe and make socket events table. 

This OpenBSM service runs concurrently with Endpoint Security service which allows Zeek-agent to collect file, process, and socket events at the same time.

### CODE REVIEW

File: components/zeekopenbsm/src/openbsmconsumer.cpp [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+        event.header.success = 1;
+      } else {
+        event.header.success = 0;
+      }
+    } else
+      event.header.success = 0;
+    break;
+  }
+  }
+}
```
Comment (bbannier) [inline components/zeekopenbsm/src/openbsmconsumer.cpp:170]: We could make this slightly less repetitive by extracting the status to expect in the switch,
```suggestion

void OpenbsmConsumer::extractReturn(IOpenbsmConsumer::Event &event,
                                    tokenstr_t tok) {
  u_char status = 0;
  switch (tok.id) {
  case AUT_RETURN32:
    status = tok.tt.ret32.status;
    break;
  case AUT_RETURN64:
    status = tok.tt.ret64.err;
    break;
  default:
    return;
  }

  int error = 0;
  if (au_bsm_to_errno(status, &error) == 0) {
    if (error == 0) {
      event.header.success = 1;
    } else {
      event.header.success = 0;
    }
  } else
    event.header.success = 0;
}

```

File: components/zeekopenbsm/src/openbsmconsumer.cpp [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+
+  switch (tok.id) {
+  case AUT_SOCKINET32: {
+    if (event.type == Event::Type::Bind) {
+      event.header.remote_address = "0.0.0.0";
+      event.header.remote_port = 0;
+      event.header.local_address = getIpFromToken(tok);
+      event.header.local_port = ntohs(tok.tt.sockinet_ex32.port);
+    } else {
+      event.header.remote_address = getIpFromToken(tok);
```
Comment (bbannier) [inline components/zeekopenbsm/src/openbsmconsumer.cpp:182]: We could `assert(event.type == Event::Type::Connect)` here.

File: components/zeekopenbsm/src/openbsmconsumer.cpp [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+  }
+}
+void OpenbsmConsumer::extractSocketInet(IOpenbsmConsumer::Event &event,
+                                        tokenstr_t tok) {
+
+  switch (tok.id) {
+  case AUT_SOCKINET32: {
+    if (event.type == Event::Type::Bind) {
+      event.header.remote_address = "0.0.0.0";
+      event.header.remote_port = 0;
```
Comment (bbannier) [inline components/zeekopenbsm/src/openbsmconsumer.cpp:178]: Does it make sense to fill these fields with dummy values? Maybe we should just leave them unset (with the types you picked: empty string for `address`, `0` for `port`).

File: components/zeekopenbsm/src/openbsmconsumer.cpp [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+      event.header.local_address = "0.0.0.0";
+      event.header.local_port = 0;
+    }
+    if (tok.tt.sockinet_ex32.family == 2) {
+      event.header.family = 2;
+    } else if (tok.tt.sockinet_ex32.family == 26) {
+      event.header.family = 10;
+    } else {
+      event.header.family = 0;
+    }
```
Comment (bbannier) [inline components/zeekopenbsm/src/openbsmconsumer.cpp:193]: This is slightly hard to follow. Can you use some existing constants for the RHS?

File: components/zeekopenbsm/src/openbsmconsumer.cpp [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+
+    break;
+  }
+  case AUT_SOCKINET128: {
```
Comment (bbannier) [inline components/zeekopenbsm/src/openbsmconsumer.cpp:197]: See comments on `case AUT_SOCKINET32` above.

File: components/zeekopenbsm/src/openbsmconsumer.cpp [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+
+  Status status;
+  Event event;
+
+  u_char *buffer = nullptr;
+
+  int reclen = au_read_rec(audit_pipe, &buffer);
+
+  if (reclen <= 0) {
+    return Status::failure("Could not openbsm fetch message");
```
Comment (bbannier) [inline components/zeekopenbsm/src/openbsmconsumer.cpp:269]: Can we include `strerror(errno)` in this error message?

File: components/zeekopenbsm/src/openbsmconsumer.cpp [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+  }
+
+  tokenstr_t tok;
+  std::vector<tokenstr_t> tokens{};
+  tokens.reserve(12);
```
Comment (bbannier) [inline components/zeekopenbsm/src/openbsmconsumer.cpp:274]: Where is this number coming from? Shouldn't this be (derived from) `reclen` instead?

File: components/zeekopenbsm/src/openbsmconsumer.cpp [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+      event_id = tok.tt.hdr64.e_type;
+      break;
+    case AUT_HEADER64_EX:
+      event_id = tok.tt.hdr64_ex.e_type;
+      break;
+    }
+    tokens.push_back(tok);
+    bytesread += tok.len;
+  }
+  if (event_filter_list.find(event_id) == event_filter_list.end()) {
```
Comment (bbannier) [inline components/zeekopenbsm/src/openbsmconsumer.cpp:299]: Doesn't this assume that the tokens we receive are ordered in a way where "interesting" tokens always preceed "uninteresting" ones? We would skip the rest of the function, even if we found some interesting ones.

Maybe I am misunderstanding what this does, but we also seem to reuse the `event_id` of the previous iteration if no case matches. Shouldn't we reset its value in a default case?

File: components/zeekopenbsm/src/openbsmconsumer.cpp [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+}
+
+Status OpenbsmConsumer::fetchEventsFromPipe() {
+
+  audit_pipe = fopen(AUDIT_PIPE_PATH, "r");
+  if (audit_pipe == nullptr) {
+    throw Status::failure("The auditpipe couldn't be opened.");
+  }
+  event_filter_list.insert(AUE_CONNECT);
+  event_filter_list.insert(AUE_BIND);
```
Comment (bbannier) [inline components/zeekopenbsm/src/openbsmconsumer.cpp:324]: Can we do this in the constructor instead? As written it can race with any reads of this variable.

File: components/zeekopenbsm/src/openbsmconsumer.cpp [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+  fd_set fdset;
+  struct timeval timeout {};
+  timeout.tv_sec = 0;
+  timeout.tv_usec = 200000;
```
Comment (bbannier) [inline components/zeekopenbsm/src/openbsmconsumer.cpp:328]:
```suggestion

  fd_set fdset;

  timeval timeout;
  timeout.tv_sec = 0;
  timeout.tv_usec = 200'000;
```

File: components/zeekopenbsm/src/openbsmconsumer.cpp [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+    FD_SET(fileno(audit_pipe), &fdset);
+
+    int rc = select(FD_SETSIZE, &fdset, nullptr, nullptr, &timeout);
+
+    if (rc == 0) {
+      continue;
+    }
+    if (rc < 0) {
+      if (errno != EINTR) {
+        return Status::failure("Auditpipe cannot be read");
```
Comment (bbannier) [inline components/zeekopenbsm/src/openbsmconsumer.cpp:341]: Let's include `strerror(errno)` in the error message.

File: components/zeekopenbsm/src/openbsmconsumer.cpp [hunk 12]
```diff
@@ -0,0 +1,363 @@
...
+      }
+      continue;
+    }
+    parseMessage();
```
Comment (bbannier) [inline components/zeekopenbsm/src/openbsmconsumer.cpp:345]: We drop any error messages from this function on the floor. We should the very least log them, alternatively we could also transport them through (a modified) `event_list` to consumers.

File: components/zeekopenbsm/src/openbsmconsumer.cpp [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+  }
+  return Status::success();
```
Comment (bbannier) [inline components/zeekopenbsm/src/openbsmconsumer.cpp:347]: We never collect this return value. We could either transport it to the caller by having him pass a promise we set, or by just logging. In any case, this function could be `void`.

File: components/zeekopenbsm/src/openbsmconsumer.cpp [hunk 14]
```diff
@@ -0,0 +1,363 @@
...
+}
+
+Status IOpenbsmConsumer::create(Ref &obj, IZeekLogger &logger,
+                                IZeekConfiguration &configuration) {
+  try {
+    obj.reset(new OpenbsmConsumer(logger, configuration));
+    return Status::success();
+
+  } catch (const std::bad_alloc &) {
+    return Status::failure("Memory allocation failure");
```
Comment (bbannier) [inline components/zeekopenbsm/src/openbsmconsumer.cpp:357]: Personally I find handling out-of-memory errors superfluous since we allocate in a lot of other places in the code where we never handle failures (e.g., we use `string`). I understand this is here for consistency with the rest of the code.

File: components/zeekopenbsm/tests/openbsmconsumer.cpp [hunk 15]
```diff
@@ -0,0 +1,103 @@
+#include "openbsmconsumer.h"
```
Comment (bbannier) [inline components/zeekopenbsm/tests/openbsmconsumer.cpp:1]: Could you include some tests here which exercise an actual `OpenbsmConsumer` instance?

Maybe it would make sense to change to a more end-to-end strategy here instead. For that you could change the constructor of `OpenbsmConsumer` so one can optionally pass in an alternative audit pipe path (defaulting to the "real" value). You could then inject events into a test pipe. That would also give us a chance to convince ourselves that we haven't introduced deadlocks, and to more exhaustively exercise the `extract*` functions where the current tests only cover a small subset.

File: tables/openbsm/src/openbsmservice.cpp [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+
+  try {
+    auto ptr =
+        new OpenbsmServiceFactory(virtual_database, configuration, logger);
+    obj.reset(ptr);
+
+    return Status::success();
+
+  } catch (const std::bad_alloc &) {
+    return Status::failure("Memory allocation failure");
```
Comment (bbannier) [inline tables/openbsm/src/openbsmservice.cpp:131]: Ditto.

File: tables/openbsm/src/openbsmservice.cpp [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+  obj.reset();
+
+  try {
+    obj.reset(
+        new OpenbsmService(d->virtual_database, d->configuration, d->logger));
+
+    return Status::success();
+
+  } catch (const std::bad_alloc &) {
+    return Status::failure("Memory allocation failure");
```
Comment (bbannier) [inline tables/openbsm/src/openbsmservice.cpp:152]: Ditto.

**Review by bbannier — CHANGES_REQUESTED** [review]

File: components/zeekopenbsm/src/openbsmconsumer.cpp [hunk 18]
```diff
@@ -0,0 +1,363 @@
...
+
+  std::atomic_bool terminate_producer{false};
+};
+
+OpenbsmConsumer::OpenbsmConsumer(IZeekLogger &logger,
+                                 IZeekConfiguration &configuration)
+    : d(new PrivateData(logger, configuration)) {
+
+  // start fetching events from pipe in a separate thread
+  std::thread{&OpenbsmConsumer::fetchEventsFromPipe, this}.detach();
```
[the remark this answers is not in the document]
Comment (Wajihulhassan) [reply components/zeekopenbsm/src/openbsmconsumer.cpp:36]: I did not use `join` because there was no reason for me to wait for consumer thread. As soon as the destructor for the producer is called, the consumer thread will be stopped. As of now, we don't have multiple Openbsm producers. We just have one Openbsm producer and one consumer; thus, leaking will never happen.

File: components/zeekopenbsm/src/openbsmconsumer.cpp (hunk 7 above)
Comment (Wajihulhassan) [reply components/zeekopenbsm/src/openbsmconsumer.cpp:270]: Basically, we need an upper bound on the number of token strings that can appear per socket bind/connect event. I could not find any documentation that can give me an upper bound value. Further, I don't think we can derive that number from `reclen` because `reclen` represents total bytes in all the token strings per event rather than the total number of token strings per event. So, I looked into osquery code and they were using 12 as an upper bound.

File: components/zeekopenbsm/src/openbsmconsumer.cpp (hunk 8 above)
Comment (Wajihulhassan) [reply components/zeekopenbsm/src/openbsmconsumer.cpp:295]: So it works like this: each event (e.g. socket connect) generates multiple token strings. Each generated token string has a type that is defined by `type.id`. HEADER is a must token type that appears only once per event. In this case, `event_id` will only be assigned once in the loop since there is only one token string with HEADER type.
I hope this anwsers your first question as well. Let me know if it is still unclear.

File: components/zeekopenbsm/src/openbsmconsumer.h [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+  /// \brief Returns a list of processed events
+  /// \param event_list Where the event list is stored
+  /// \return A Status object
+  Status getEvents(EventList &event_list) override;
+
+private:
+  struct PrivateData;
+  std::unique_ptr<PrivateData> d;
+  // audit pipe handler
+  FILE *audit_pipe{nullptr};
```
[the remark this answers is not in the document]
Comment (Wajihulhassan) [reply components/zeekopenbsm/src/openbsmconsumer.h:23]: This was a bug. It should've been doxygen comment

File: components/zeekopenbsm/src/openbsmconsumer.h [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+  OpenbsmConsumer(IZeekLogger &logger, IZeekConfiguration &configuration);
+
+  /// \brief extract header from openbsm token and populate event
+  static void extractHeader(Event &event, tokenstr_t tok);
+  /// \brief extract subject from openbsm token and populate event
+  static void extractSubject(Event &event, tokenstr_t tok);
+  /// \brief extract return from openbsm token and populate event
+  static void extractReturn(Event &event, tokenstr_t tok);
+  /// \brief extract socket-inet from openbsm token and populate event
+  static void extractSocketInet(Event &event, tokenstr_t tok);
```
[the remark this answers is not in the document]
Comment (Wajihulhassan) [reply components/zeekopenbsm/src/openbsmconsumer.h:38]: No specific reason. Can this create some issue? What do you suggest?

File: components/zeekopenbsm/tests/openbsmconsumer.cpp (hunk 15 above)
Comment (Wajihulhassan) [reply components/zeekopenbsm/tests/openbsmconsumer.cpp:1]: I can add a test for e2e testing `OpenbsmConsumer` but that requires a lot of changes. I prefer to make a separate PR for that test. Further, as of now we only have one consumer and one producer running, there are fewer chances that we introduced some deadlocks.

File: tables/openbsm/include/zeek/openbsmservicefactory.h [hunk 21]
```diff
@@ -0,0 +1,18 @@
...
+#include <zeek/izeekservicemanager.h>
+
+namespace zeek {
+/// \brief Factory method for the EndpointSecurityServiceFactory object
+/// \param service_manager An initialized service manager
+/// \param virtual_database The database where the EndpointSecurity tables
+///                         are registered
+/// \param configuration An initialized configuration object
+/// \param logger An initialized logger object
+Status registerOpenbsmServiceFactory(IZeekServiceManager &service_manager,
```
[the remark this answers is not in the document]
Comment (Wajihulhassan) [reply tables/openbsm/include/zeek/openbsmservicefactory.h:13]: Do we explicitly need to add `zeek/status.h` here if already included file `zeek/izeekconfiguration.h` has `zeek/status.h` in its includes?

File: tables/openbsm/include/zeek/openbsmservicefactory.h [hunk 22]
```diff
@@ -0,0 +1,18 @@
...
+                                     IVirtualDatabase &virtual_database,
```
[the remark this answers is not in the document]
Comment (Wajihulhassan) [reply tables/openbsm/include/zeek/openbsmservicefactory.h:14]: same as above.

File: tables/openbsm/include/zeek/openbsmservicefactory.h [hunk 23]
```diff
@@ -0,0 +1,18 @@
...
+                                     IZeekConfiguration &configuration,
+                                     IZeekLogger &logger);
```
[the remark this answers is not in the document]
Comment (Wajihulhassan) [reply tables/openbsm/include/zeek/openbsmservicefactory.h:16]: same as above.

File: tables/openbsm/src/openbsmservice.cpp (hunk 16 above)
Comment (Wajihulhassan) [reply tables/openbsm/src/openbsmservice.cpp:131]: As I mentioned above, do we need explicit include here?

File: tables/openbsm/src/openbsmservice.cpp (hunk 17 above)
Comment (Wajihulhassan) [reply tables/openbsm/src/openbsmservice.cpp:152]: same as above.

File: components/zeekopenbsm/src/openbsmconsumer.h (hunk 20 above)
[the remark this answers is not in the document]
Comment (bbannier) [reply components/zeekopenbsm/src/openbsmconsumer.h]: Out parameters are harder to understand than simply returning a value since they e.g., force callers to default-construct a value which is only set up later. This can make it e.g., less obvious whether a value is set on all paths.

I'd suggest changing all the functions here to take a `tokenstr_t` and to return an `Event` by value. `populateSocketEvent` take a `const vector<tokenstr_t>&` should return an `Event` as well (the status is not needed).

File: components/zeekopenbsm/src/openbsmconsumer.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (bbannier) [reply components/zeekopenbsm/src/openbsmconsumer.cpp:39]: Not waiting here can obscure any issues in the shutdown of the threads (which might be introduced inadvertently a some later time). Let's make sure the thread actually has terminated by joining it.

File: tables/openbsm/include/zeek/openbsmservicefactory.h (hunk 21 above)
[the remark this answers is not in the document]
Comment (bbannier) [reply tables/openbsm/include/zeek/openbsmservicefactory.h:13]: I'd argue we should include it to make this file independent of other header files (even ones it includes). https://google.github.io/styleguide/cppguide.html#Self_contained_Headers

File: components/zeekopenbsm/src/openbsmconsumer.cpp (hunk 7 above)
Comment (bbannier) [reply components/zeekopenbsm/src/openbsmconsumer.cpp:259]: Let's either add a comment capturing how we arrive at this number, or get rid of the `reserve`.

File: components/zeekopenbsm/src/openbsmconsumer.cpp (hunk 8 above)
Comment (bbannier) [reply components/zeekopenbsm/src/openbsmconsumer.cpp:287]: Thanks, that helps. Can we work on making the implementation more closely model what you explained? We could e.g., log a (fatal?) error if `au_fetch_tok` fails, move the storing of `tok` in `tokens` right after the fetch since they belong together, make `event_id` a `optional<u_int16_t>` and `assert(event_id)` before using it, and expand the comment for the early return with something capturing the guaranteed semantics.

File: components/zeekopenbsm/tests/openbsmconsumer.cpp (hunk 15 above)
Comment (bbannier) [reply components/zeekopenbsm/tests/openbsmconsumer.cpp:1]: We should really include a test of this functionality with this PR, currently we have almost no coverage. How did you test this?

File: components/zeekopenbsm/src/openbsmconsumer.cpp (hunk 14 above)
Comment (Wajihulhassan) [reply components/zeekopenbsm/src/openbsmconsumer.cpp:347]: Yes, you are correct. I am doing this just to be consistent with the current code style.

File: components/zeekopenbsm/src/openbsmconsumer.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (Wajihulhassan) [reply components/zeekopenbsm/src/openbsmconsumer.cpp:39]: Alright. That makes sense. I'll add `join` in the destructor.

File: components/zeekopenbsm/src/openbsmconsumer.cpp (hunk 7 above)
Comment (Wajihulhassan) [reply components/zeekopenbsm/src/openbsmconsumer.cpp:259]: I'll get rid of the `reserve` because I am not sure how memory saving it buys us.

File: components/zeekopenbsm/src/openbsmconsumer.cpp (hunk 12 above)
Comment (Wajihulhassan) [reply components/zeekopenbsm/src/openbsmconsumer.cpp:335]: Now `parseMessage` function logs any failure that happens in that function.

File: components/zeekopenbsm/src/openbsmconsumer.h (hunk 20 above)
[the remark this answers is not in the document]
Comment (Wajihulhassan) [reply components/zeekopenbsm/src/openbsmconsumer.h]: In the latest commit, these functions return `Status` object to show if the extraction was successful or not so that we can log in a failure scenario. Let me know what you think.

File: components/zeekopenbsm/src/openbsmconsumer.cpp (hunk 8 above)
Comment (Wajihulhassan) [reply components/zeekopenbsm/src/openbsmconsumer.cpp]: I added your suggestion in the lastest commit.

File: components/zeekopenbsm/tests/openbsmconsumer.cpp (hunk 15 above)
Comment (Wajihulhassan) [reply components/zeekopenbsm/tests/openbsmconsumer.cpp:1]: The workflow of producer in openbsmconsumer.cpp is as follows:
`fetchRecordsFromAuditPipe` -> `parseRecordIntoTokens` -> `populateEventFromTokens` -> `extract*`.
Now, we have tests for `populateEventFromTokens` function which basically calls all the `extract*` methods to return `Event` object. Thus, our coverage is pretty good already.
To write a test that starts from `fetchRecordsFromAuditPipe` function is not that easy because you need to make sure in this new test that whatever you write to the pipe must be readable in `parseRecordIntoTokens`  using `au_read_rec` function. I cannot write some arbitrary text into the pipe and then read it using `au_read_rec`. 

Besides that, the grunt of work is done in `populateEventFromTokens` function, which we are already testing while `fetchRecordsFromAuditPipe` and `parseRecordIntoToken` functions just read records from the pipe and then tokenize them, respectively.

File: components/zeekopenbsm/src/openbsmconsumer.cpp [hunk 24]
```diff
@@ -0,0 +1,361 @@
...
+      return;
+    }
+    bytesread += tok.len;
+  }
+
+  assert(event_id);
+
+  if (subscribed_event_ids.find(event_id.value()) ==
+      subscribed_event_ids.end()) {
+    // we will not store current record in the event_list
```
Comment (bbannier) [inline components/zeekopenbsm/src/openbsmconsumer.cpp:296]: Let's document _why_ instead of _how_, e.g.,
```suggestion
    // Skip events not subscribed to.
```

File: components/zeekopenbsm/src/openbsm_utils.cpp [hunk 25]
```diff
@@ -0,0 +1,33 @@
...
+  return Status::failure("Cannot parse IP address from given token due to "
+                         "unhandled socket family");
+}
+
+Status getPathFromPid(int pid, std::string &path) {
+  char pathbuf[PROC_PIDPATHINFO_MAXSIZE] = {0};
+
+  int ret = proc_pidpath(pid, pathbuf, sizeof(pathbuf));
+  if (ret > 0) {
+    path = std::string(pathbuf);
```
Comment (bbannier) [inline components/zeekopenbsm/src/openbsm_utils.cpp:27]: Small nit since I see it here:
```suggestion
    ip_addr = inet_ntop(AF_INET, &ipv4, ip_str, INET6_ADDRSTRLEN);
    return Status::success();
  } else if (sock.family == 26) {
    struct in6_addr ipv6 {};
    memcpy(&ipv6, sock.addr, sizeof(ipv6));
    ip_addr = inet_ntop(AF_INET6, &ipv6, ip_str, INET6_ADDRSTRLEN);
    return Status::success();
  }
  return Status::failure("Cannot parse IP address from given token due to "
                         "unhandled socket family");
}

Status getPathFromPid(int pid, std::string &path) {
  char pathbuf[PROC_PIDPATHINFO_MAXSIZE] = {0};

  int ret = proc_pidpath(pid, pathbuf, sizeof(pathbuf));
  if (ret > 0) {
    path = pathbuf;
```

**Review by bbannier — COMMENTED** [review]
Looks good to me but for two additional nits and the outstanding test.

File: components/zeekopenbsm/tests/openbsmconsumer.cpp (hunk 15 above)
Comment (Wajihulhassan) [reply components/zeekopenbsm/tests/openbsmconsumer.cpp:1]: @bbannier To elaborate more on this, like all the other tests we have for Zeek Agent, we mock a log message from the underlying service (e.g., auditd, openbsm) _without_ running the service and then we have tests that checks the validity of our parsing functions. So I used the same pattern here. 
If you still think we should write a test that generates the log messages by running the actual service then could you give me some advice on how should I implement it.

Comment (bbannier) [reply components/zeekopenbsm/tests/openbsmconsumer.cpp:1]: > Besides that, the grunt of work is done in `populateEventFromTokens` function, which we are already testing ...

I had looked at coverage information from running the tests, and for some reason that reported no coverage where in the code there clearly is coverage. This seems good enough to me. Sorry for the hold-up because of that.

**Review by bbannier — APPROVED** [review]

**Comment by bbannier** [comment]
Looks good to me and ready for merge once CI passes.

**Comment by bbannier** [comment]
@Wajihulhassan, can you rebase this on the current `master` and check CI? Like I wrote, if CI passes, you could then go ahead and merge this.
